### PR TITLE
fix of applying default annotations on type arguments of supertypes

### PR DESCRIPTION
**Pull request by CharlesZ-Chen**
Originally, `SupertypeFinder#supertypesFromElement(...)` doesn't apply default computed annotations on `superclass` of `typeElement`, which will cause the `typeArguments` on `superclass` missing annotations in some situations.

To fix this problem, in `SupertypeFinder#supertypesFromElement(...)`, I add a code block to apply default computed annotations if the `typeElement` doesn't has any `rawTypeAttributes`, which means this `typeElement` is from bytecode that has no annotations information stored in `rawTypeAttributes`.

### CODE REVIEW

**Comment by wmdietl** [comment]
Can you see whether you can write a test case that illustrates the issue?
Maybe in the style of https://github.com/opprop/checker-framework/blob/master/framework/tests/test-lubglb/lubglb/LubGlbChecker.java ?

**Comment by wmdietl** [comment]
Also note that the Travis tests are failing, so this fix might not work correctly.

**Comment by CharlesZ-Chen** [comment]
I re-start the travis test for `opprop/master`, and that one is also failed the travis test. Maybe this is because our repository is a bit out-date compare to the `typetools/master`?

File: framework/src/org/checkerframework/framework/type/SupertypeFinder.java [hunk 1]
```diff
@@ -215,7 +217,15 @@
                 AnnotatedDeclaredType dt =
                         (AnnotatedDeclaredType) atypeFactory.toAnnotatedType(superClass, false);
                 supertypes.add(dt);
-
+                // CZC: apply defaulting to the type arguments if the rawTypeAttributes of this typeElement is empty
+                if (((ClassSymbol) typeElement).getRawTypeAttributes().isEmpty()){
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/SupertypeFinder.java]: What if there are two type arguments, but the RawTypeAttributes only contains a single type annotation.
Then the resulting type will be annotated only partially.
What happens if you remove this if?

File: framework/src/org/checkerframework/framework/type/SupertypeFinder.java [hunk 2]
```diff
@@ -215,7 +215,13 @@
                 AnnotatedDeclaredType dt =
                         (AnnotatedDeclaredType) atypeFactory.toAnnotatedType(superClass, false);
                 supertypes.add(dt);
-
+                // CZC: always try to apply defaulting to the type arguments
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/SupertypeFinder.java]: Remove the `CZC` part and you're not trying to do anything - you are applying the defaults.
So this comment should be something like `// add computed type annotations to the type arguments of the super type`

File: framework/src/org/checkerframework/framework/type/SupertypeFinder.java [hunk 3]
```diff
@@ -215,7 +215,13 @@
...
+                final List<? extends TypeMirror> typeArgs = superClass.getTypeArguments();
+                final List<AnnotatedTypeMirror> annotatedTypeArgs = dt.getTypeArguments();
+                for (int i = 0; i < typeArgs.size(); i++) {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/SupertypeFinder.java]: I personally like `++i` better for the increments.

**Comment by wmdietl** [comment]
Ping. Let's wrap this up and make a PR against upstream.

**Comment by wmdietl** [comment]
@CharlesZ-Chen Ping :-)
29 commits for 7 lines of changes. We should wrap this up and make a PR against typetools. If you do this on Wednesday, we might still get it into the next release.

**Comment by CharlesZ-Chen** [comment]
Sorry for the late response. Sure, will clean up it.

**Comment by wmdietl** [comment]
Hi @CharlesZ-Chen ... what are our next steps on this PR?

**Comment by CharlesZ-Chen** [comment]
Hi Prof. Dietl,

I remembered the changes this PR broke some implicit assumptions made in Checker-Framework, which cause we cannot easily fix the problem in this PR.

Could you give me some time to re-pick the context of this PR?  I think one valuable next step would be  documentation --- which implicit assumption this PR broke, why the problem in this PR is hard to fix, and what are possible solutions for fixing this problem.

Best,

Charles

**Comment by wmdietl** [comment]
Did you get a chance to look into this?

**Comment by CharlesZ-Chen** [comment]
Still not... Will look this now.

**Comment by CharlesZ-Chen** [comment]
Some useful discussion reference:

https://github.com/pascaliUWat/checker-framework-inference/issues/15

**Comment by CharlesZ-Chen** [comment]
## Investigation result:

I think this problem actually reflect a design conflict in type-checking (checker-framework) and type-inference (checker-framework-inference).

For illustrating, using below code as an example:

```java

class Bar<T> {
  T t;
}

class Foo extends Bar<String> {

}
```

Given an instance of `Foo`, named `foo`. Consider evaluating/inferring the type of expression element of `foo.t`.

In type-checking, checker-framework will compute the type of `foo.t` by computing what is the type of `t` when `t` as a member of reciever `foo`. Thus, as long as we can correctly get the type of `foo`, and the type of `t`, we can correctly compute the type of `foo.t`. In other words, we don't need to query any "non-persistent" (not written in source code) annotations on the type argument `String` of the super type of `Foo`. And that is the `type-check` logic --- do not pre-cache annotations on type arguments of super type of a class, and computing it by `asMemeberOf` when needed.

However, in inference-mode, this logic becomes itchy to adapt with. When we want to infer the type of `foo.t`, we first build a logical variable ("VarAnno with id") represents `foo.t`, and build related constraints, then let solver solve a value for it. Ideally (when combineConstraint doesn't involved), the logical variable represent `foo.t` should be the same as the logical variable represent the type argument `String` of the super type of `Foo`. However, as CF do not pre-cache "non-persistent" annotations on type arguments of super type of a class, we lost the access to the logical variable on supertype type argument `String` of `Foo` after the class tree of `Foo` has been visited. Thus, currently when we try to access the VarAnno on `foo.t`, CF-Inference will crash and complain cannot find this VarAnno (reflected as empty set issue, create constraint on null slot...). 

In my previous ugly fix in this PR, I try to pre-cache the "non-persistent" annotations on supertype type arguments, but unfortunately this seems break `type-check` logic for some checkers. (Reasons still un-clear, but I guess it is because some conflicts on the default result for supertype type argument and the computed result by asMemeberOf). Also, this fix is a bit ugly (not easy to understand in the code, pre-cache all supertypes annotations, query annotations depends on a cache...).

In conclusion, I don't think the fix in this PR is a proper solution, and this problem need more careful consideration about how to solve it properly.

## Possible solution

Always use combineConstraint(receiver, target) to express the logical variable for a field/methodInvocation through a receiver (e.g. `foo.t`).  Currently, I think this is the most proper and flexible solution of above problem. 

Pros:

- Consist with `type-check` logic, don't pre-cache annotations, build a combineConstraint when needed.

- Flexible, if a type system has specific view point adaptation rule, using combineConstraint could be more flexible to support that.

Cons:

 - Currently it seems we don't always enable CombineConstraints for all inference type systems. Not sure what will happens if we enable it as a general constraint.

 - Increase complexity. CombineConstraint is a bit more complex than general constraints, need to carefully think how to implement/adapt this combineConstraint to support above use case.

**Comment by topnessman** [comment]
I think this crash is only specific to CFI. 
According to my investigation, on this supertype's type argument problem, CF and CFI begin to diverge from `postDirectSuperType()` method.
In CF, its implementation is to use `treeAnnotator` and `defaultApplier` to visit the type. Supertype's type argument falls into the `OTHERWISE` type use location category in `defaultApplier` so the super type's argument will be applied default annotation in hierarchy. 
But on inference side, we don't need that default mechanism. Instead we either insert a variable annotation to the supertype's type argument if there is corresponding or apply default to the type argument if the supertype is from bytecode. But on the current implementation of `postDirectSuperType()`, we don't have such logic: we are only copying main annotation from 'type' to 'supertype':

https://github.com/opprop/checker-framework-inference/blob/32a19a3bc96dfa717b00d2895c2f082838793243/src/checkers/inference/InferenceAnnotatedTypeFactory.java#L261

I think the direction might be addComputedAnnotations to the type. So supertype's type argument is correctly inserted annotation.
I proposed an ugly fix: https://gist.github.com/topnessman/9d43341afdd7ac3ba1251f85619eb4b6
And supertype's type argument correctly has var annotation on it.

**Comment by CharlesZ-Chen** [comment]
Try if https://github.com/typetools/checker-framework/pull/1629 fixed the issue.

**Comment by wmdietl** [comment]
@xingweitian Can you have a look at the last few comments in this PR and see whether anything remains to be done?
It sounds like we can discard these changes, but please have a look first.

**Comment by xingweitian** [comment]
> @xingweitian Can you have a look at the last few comments in this PR and see whether anything remains to be done?
> It sounds like we can discard these changes, but please have a look first.

No problem, I will do some investigation.

**Comment by xingweitian** [comment]
@wmdietl It looks like typetools#1629 has resolved this issue, I also run the example code posted by Charles:
```
class Bar<T> {
  T t;
}

class Foo extends Bar<String> {

}
```
and CFI did not crash.

**Pull request closed without merging**